src/kex.c

```diff
@@ -133,7 +135,9 @@ static int diffie_hellman_sha1(LIBSSH2_SESSION *session,
         memset(&exchange_state->req_state, 0, sizeof(packet_require_state_t));
 
         /* Generate x and e */
-        _libssh2_bn_rand(exchange_state->x, group_order, 0, -1);
+        /* FIXME: too expensive, should not need more than 2*symmetric_bits */
```

Comment:
I believe that this property only holds where p is a safe prime. This should be fine for kex_method_diffie_helman_group14_sha1 and kex_method_diffie_helman_group1_sha1 but in kex_method_diffie_helman_group_exchange_sha1 and kex_method_diffie_helman_group_exchange_sha256 we're negotiating a "random(ish)" prime. Given that generating a safe prime is pretty time intensive there's a very good chance that this won't use safe primes often, so using short exponents with a p that isn't a Sophie Germaine prime seems dangerous.

I'm not a cryptographer by any stretch of the definition but http://link.springer.com/chapter/10.1007%2F3-540-68339-9_29 does make a case that this is risky.